Framework/Geometry/src/Instrument/StructuredDetector.cpp

```diff
@@ -0,0 +1,701 @@
+                                    const std::vector<double> &y, int idstart,
+                                    bool idfillbyfirst_y, int idstepbyrow,
+                                    int idstep) {
+  if (m_map)
+    throw std::runtime_error("StructuredDetector::initialize() called for a "
+                             "parametrized StructuredDetector");
+
+  m_xpixels = xpixels;
+  m_ypixels = ypixels;
+
+  /// IDs start here
+  m_idstart = idstart;
+  /// IDs are filled in Y fastest
+  m_idfillbyfirst_y = idfillbyfirst_y;
+  /// Step size in ID in each row
+  m_idstepbyrow = idstepbyrow;
+  /// Step size in ID in each col
+  m_idstep = idstep;
+
+  // Some safety checks
+  if (m_xpixels <= 0)
+    throw std::invalid_argument(
+        "StructuredDetector::initialize(): xpixels should be > 0");
+  if (m_ypixels <= 0)
+    throw std::invalid_argument(
+        "StructuredDetector::initialize(): ypixels should be > 0");
+  if (x.size() != y.size())
+    throw std::invalid_argument(
+        "StructuredDetector::initialize(): x.size() should be = y.size()");
+  if (x.size() != ((m_xpixels + 1) * (m_ypixels + 1)))
+    throw std::invalid_argument("StructuredDetector::initialize(): x.size() "
+                                "should be = (xpixels+1)*(ypixels+1)");
+
+  // Store vertices
+  m_xvalues = x;
+  m_yvalues = y;
+
+  std::string name = this->getName();
+  auto minDetId = idstart;
+  auto maxDetId = idstart;
+  auto detSize = x.size();
+
+  CompAssembly *detAssembly = new CompAssembly(name);
+
+  for (auto ix = 0; ix < m_xpixels; ix++) {
+    // Create an ICompAssembly for each x-column
+    std::ostringstream oss_col;
+
+    oss_col << name << "(x=" << ix << ")";
+    CompAssembly *xColumn = new CompAssembly(oss_col.str(), this);
+
+    for (auto iy = 0; iy < m_ypixels; iy++) {
+      std::ostringstream oss;
+      oss << name << "(" << ix << "," << iy << ")";
+
+      // Calculate its id and set it.
+      auto id = this->getDetectorIDAtXY(ix, iy);
+
+      // Store min and max det IDs
+      // minimum structured detector id
+      if (id < minDetId) {
+        minDetId = id;
+      }
+      // maximum structured detector id
+      if (id > maxDetId) {
+        maxDetId = id;
+      }
+
+      // Create and store detector pixel
+      auto detector = addDetector(xColumn, oss.str(), ix, iy, id);
+      xColumn->add(detector);
+    }
+  }
+
+  m_minDetId = minDetId;
+  m_maxDetId = maxDetId;
+}
+
+/** Creates new hexahedral detector pixel at row x column y using the
+*   detector vertex values.
+* @param name :: The Pixel name identifier
+* @param x :: The pixel row
+* @param y :: The pixel column
+* @param id :: The pixel ID
+* @ returns newly created detector.
+*/
+Detector *StructuredDetector::addDetector(CompAssembly *parent,
+                                          const std::string &name, int x, int y,
+                                          int id) {
+  auto w = m_xpixels + 1;
```

Comment:
I see nothing wrong with your overall procedure here, but method is too large. The stream part could be done in a separate function.